### ISSUE TITLE
AutocompleteInput does not handle "manual" changes well

### ISSUE THREAD

**Issue by nirAlt19** [bokeh/bokeh#9605]
bokeh --version
1.4.0

python --version
Python 3.7.6

Expected: AutocompleteInput reflects any text changes made in the text input box, no matter how they got there.
Actual: When changing the value in the textbox, the values reflected in the textbox are the last change made from the completions list.

It might easily be different than what the value property of the widget actually holds. The changes do not trigger the on_change method.

I used the following example, taken from another issue.

```python
from bokeh.models.widgets.inputs import AutocompleteInput
from bokeh.plotting import curdoc

input_box = AutocompleteInput()
input_box.title = "Foo Bar"
input_box.value = "400"
input_box.completions = ["100001", "12344556", "3194567289", "209374209374"]


def update():
    print(input_box.value)


input_box.on_change('value', lambda attr, old, new: update())

curdoc().add_root(input_box)
```

The user inputs 10 and selects 100001.
The user tries options starting with 95. There aren't any.
The number 95 remains in the textbox, while the widget's value remains 100001.
The worst part is that if the user decides to re-select 100001, the option of 100001 appears, but upon selecting it, the textbox still shows 95.

The same holds if, for some reason, we would like to allow the user to input other values not in the Autocomplete list of options. The change does not trigger the attribute change, so we can't use the entered value.

**Comment by bryevdv** [comment]
> The worst part is that if the user decides to re-select 100001, the option of 100001 appears, but upon selecting it, the textbox still shows 95.

I am unable to reproduce this, a screen capture would go a long way  towards helping see exactly what you are are seeing. 

That said, offhand, I think the text box should clear to an empty value if a user tries to hit enter on a non-acceptable value. Other ideas are welcome.

**Comment by nirAlt19** [comment]
> > The worst part is that if the user decides to re-select 100001, the option of 100001 appears, but upon selecting it, the textbox still shows 95.
> 
> I am unable to reproduce this part, a screen capture would go a long way towards helping see exactly what you are are seeing.
> 
> That said, offhand, I think the text box should clear to an empty value if a user tries to hit enter on a non-acceptable value. Other ideas are welcome, though.

Sorry, it should have said "the textbox still shows 10", in case you used the substring "10" to search 100001.

I've changed the code to output the current value.

```Python
from bokeh.models.widgets.inputs import AutocompleteInput
from bokeh.models.widgets import PreText
from bokeh.layouts import column
from bokeh.plotting import curdoc



txt = PreText(text="")
input_box = AutocompleteInput()
input_box.title = "Foo Bar"
input_box.value = "400"
input_box.completions = ["100001", "12344556", "3194567289", "209374209374"]

def update():
    txt.text = f"Autocomplete value is {input_box.value}"

input_box.on_change('value', lambda attr, old, new: update())

curdoc().add_root(column(input_box, txt))```

![Untitled](https://user-images.githubusercontent.com/51610171/73120490-4d5fbf00-3f77-11ea-9312-bfccbd06ba09.jpg)

![Untitled](https://user-images.githubusercontent.com/51610171/73120498-74b68c00-3f77-11ea-9144-0262872d176a.jpg)

After using "10" to find and select 100001, this is the result:
![Untitled](https://user-images.githubusercontent.com/51610171/73120504-8ef06a00-3f77-11ea-9856-3a1bfe567147.jpg)

Thanks for the response.

**Comment by nirAlt19** [comment]
> That said, offhand, I think the text box should clear to an empty value if a user tries to hit enter on a non-acceptable value. Other ideas are welcome, though.

I think there is a room to customize this option - allow the developer to decide whether the user can enter any text to change the value, clear the box or reverting back to the last value. But this must be reflected somewhere in the attributes and trigger a catch-able on_change...

**Comment by ltalirz** [comment]
I confirm the issue observed by @nirAlt19 , namely that in certain cases, selections from the completion list are not reflected in the "value" attribute (as mentioned previously already in https://github.com/bokeh/bokeh/issues/6121#issuecomment-325179238 )

I also second the feature request in the last comment https://github.com/bokeh/bokeh/issues/9605#issuecomment-578399191, and have opened a separate issue for it: https://github.com/bokeh/bokeh/issues/10295

**Comment by ltalirz** [comment]
After some more testing, the only case where the value of the input field is not correctly updated, is if the same completion value is selected twice:

![autocompleteinput](https://user-images.githubusercontent.com/1053245/87298435-7c102880-c50a-11ea-840e-80df95245ac3.gif)

**Comment by ltalirz** [comment]
Being new to TS, I thought I start looking into fixing this issue before tackling https://github.com/bokeh/bokeh/issues/10295 since at least this one should be limited to the TS/JS side of things -- but I came across an issue that I'm having difficulties to track down: 

Looking through the code, I came across this function
https://github.com/bokeh/bokeh/blob/658f80d22fe9900a08435bb51cdf59b0900028a6/bokehjs/src/lib/models/widgets/autocomplete_input.ts#L85-L91
which would seem to be responsible for handling clicks on the autocomplete dropdown menu.
https://github.com/bokeh/bokeh/blob/658f80d22fe9900a08435bb51cdf59b0900028a6/bokehjs/src/lib/models/widgets/autocomplete_input.ts#L34


However, after a couple of unsuccessful trials to "fix" the code, it turns out that this function is *not* always triggered when clicking on the dropdown menu.

A symptom of this seems to be that, usually, the dropdown menu already disappears on `mousedown` rather than on a complete click (and in that case, the function `_menu_click` is not triggered).

In *some* cases (~roughly every 10th time; I don't know how to reliably reproduce it), the dropdown menu does *not* disappear on mousedown (e.g. allowing me to select text in the dropdown menu by dragging the mouse). In that case `_menu_click` is triggered (the menu will still disappear when completing the click with mouseup).

Does anybody have an idea what could be the problem here?
Is there an additional "mousedown" listener that is taking precedence most of the time?


## How to reproduce
To reproduce, copy the `example.py` code from [above](https://github.com/bokeh/bokeh/issues/9605#issue-555013722), and in a development environment run
```
BOKEH_MINIFIED=no BOKEH_RESOURCES=inline bokeh serve example.py  --show
```
Then set a breakpoint inside the `_menu_click` function and start playing around.

**Comment by redhog** [comment]
Workaround:

```
input = bokeh.models.AutocompleteInput(
    completions="100001", "12344556", "3194567289", "209374209374"],
    value="100001", title='test')
def fix_value(attr, old, new):
    input.value_input = input.value
input.on_change("value", fix_value)
```

Then anywhere you'd normally use `input.value`, instead use `input.value_input`.

`.value_input` is updated with every new character typed, but _not_ when the user selects from the drop down. `fix_value()` fixes this defect, so that `.value_input` always contain the latest user entry, no matter how it got there.

Regarding a permanent solution, I'd argue that `.value` should keep its current semantics, while `.value_input` should change semantics to incorporate what the above `fix_value` function does.

**Comment by ltalirz** [comment]
> `.value` should keep its current semantics, while `.value_input` should change semantics to incorporate what the above `fix_value` function does.

Makes sense - `input_el.value` *should* be updated when selecting a value from the dropdown.
(somehow it did seem to work already when the `model.value` was changed at the same time, but it should also update when the `model.value` remains unchanged). 

Would you be able to make a PR?

**Comment by redhog** [comment]
I'm sorry, but my understanding of the javascript parts of bokeh is a bit too hazy to try that (I assume a real fix would do the update in JS, not like my workaround)...

**Comment by solstag** [comment]
> After some more testing, the only case where the value of the input field is not correctly updated, is if the same completion value is selected twice in a row:

Ni! This has been an issue for me. I think it makes sense to update 'value' on all cases, be it selecting from the menu or pressing 'return'. The developer can keep track of changes of same value if they need to ignore those. But the other way around ain't true. That is, if I need to act on changes of same value I need bokeh to tell me about them.

I've bee working around this by setting 'value' back to "" at the end of my callbacks (and exiting at the beginning if that is the case). But this causes the entered text to disappear upon selection, which is not very nice for the user.
